### PR TITLE
[https://nvbugs/6384633][fix] Fix NCCL Shutdown

### PR DESCRIPTION
**Pull request by belgarten-nv**
## Summary by CodeRabbit

* **New Features**
  * Added an explicit process-group shutdown function for cleaner runtime cleanup.
  * Exposed the shutdown action to Python, so it can be called directly from scripts.
  * Automatic shutdown now runs when the Python interpreter exits, helping ensure resources are released reliably.

## Description

Add proper shutdown for NCCL process groups. This change releases references to `pg_world` and `pg_local` such that they can be destroyed before CUDA teardown. Addresses [nvbugs/6384633](https://nvbugs/6384633), [nvbugs/6398197](https://nvbugs/6398197), and [nvbugs/6398198](https://nvbugs/6398198). 

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by tongyuantongyu — APPROVED** [review]

**Review by NVShreyas — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/visual_gen/mapping.py [hunk 1]
```diff
@@ -243,9 +247,14 @@ def setup_communicators(self):
...
         from tensorrt_llm._utils import torch_pybind11_abi
-        from tensorrt_llm.bindings.internal.process_group import init_pg
+        from tensorrt_llm.bindings.internal.process_group import init_pg, shutdown_pg
 
         init_pg(dist.group.WORLD, self.local_comm, torch_pybind11_abi())
+        if not VisualGenMapping._shutdown_pg_registered:
+            import atexit
+
+            atexit.register(shutdown_pg)
+            VisualGenMapping._shutdown_pg_registered = True
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/mapping.py:257]: TorchDist calls the same `init_pg` (`communicator.py:810`) but gets no shutdown registration — the LLM/Ray path keeps the same dangling refs this PR fixes. Should the `atexit` registration live next to `init_pg` (binding wrapper or a shared util) so every caller is covered? That would also remove the class-level latch here — double-registering `shutdown_pg` is harmless since `reset()` is idempotent.

File: cpp/tensorrt_llm/runtime/utils/pgUtils.cpp [hunk 2]
```diff
@@ -41,4 +41,10 @@ void init_pg(c10::intrusive_ptr<c10d::ProcessGroup> const& process_group_world,
     pg_local = process_group_local;
 }
 
+void shutdown_pg()
+{
+    pg_world.reset();
+    pg_local.reset();
+}
```
Comment (zhenhuaw-me) [inline cpp/tensorrt_llm/runtime/utils/pgUtils.cpp:48]: This reset is unsynchronized with the external-launch worker: `VisualGen._atexit_shutdown` gives up joining the daemon worker after `WORKER_TIMEOUT=2s`, so this atexit can null the PGs while a collective is still running through `get_world_pg()`. Does it make sense to call `shutdown_pg` from the worker's teardown path (after collectives stop, alongside `destroy_process_group`) rather than a bare atexit?

Comment (belgarten-nv) [reply cpp/tensorrt_llm/runtime/utils/pgUtils.cpp:48]: These are just refcount decrements on the C++ globals — they don't destroy the underlying PG. Anything actively holding a ref (torch.dist, the device mesh, per-op comm handles) keeps it alive; in-flight collectives are unaffected. The only paths that could dereference a nulled global are hot-path callers that skip the local copy, and reaching them requires the worker to be mid-inference *after* WORKER_TIMEOUT already expired — at which point graceful shutdown has been abandoned and the daemon thread is about to be killed by interpreter finalization regardless.

Moving `shutdown_pg` into the worker's teardown path regresses the actual bug: if the worker doesn't reach teardown (crash, stuck collective, SIGTERM), the C++ globals leak and `~ProcessGroupNCCL` fires from the static destructor after CUDA has finalized — the SIGABRT repro in `dev/bug.py`.

File: tensorrt_llm/_torch/visual_gen/mapping.py (hunk 1 above)
Comment (belgarten-nv) [reply tensorrt_llm/_torch/visual_gen/mapping.py:257]: An earlier version of this PR did have that code path (I found the best way to directly call `atexit` at import time of `tensorrt_llm.bindings.internal.process_group`, which you can do via nanobind in the cpp source). However, I had many [CI failures](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/blue/organizations/jenkins/LLM%2Fmain%2FL0_MergeRequest_PR/detail/L0_MergeRequest_PR/46002/pipeline) I was unable to diagnose. The non-TRTLLM path that calls into `init_pg` is the Ray path, which is distinct from the VisualGen executor path, and it seems this bug has not previously manifested there. Thus I chose to just scope to VisualGen to fix the triggering bug. If this issue appears on the LLM path, then the infrastructure to fix is already set up for them.

**Comment by belgarten-nv** [comment]
/bot skip --comment "CI has already passed"
